Fix last checkpoint finding in filtered files with correct extension

yassersouri opened the pull request:
## What does this PR do?

Simple change. The last checkpoint candidate finding code, would look at all the files with "last" in their name.
But we could use a better filtering mechanism to include files that have "last" in their name and have the correct extension.

(since the fix is very simple, I didn't create an issue for it)

I don't think this breaks anything.

## What was my issue?
After the training finished, I was saving some other files in the directory with "last" and "best" in their names. These files were tsv, text or other kind of files. 
After the save, the testing using ` ckpt_path="last"` would stop working and I noticed in the logs that one of the files that I had created was being loaded instead of the correct "last.ckpt" file.
I have tested locally and with this simple change, this issue for me is resolved.

carmocca commented on src/lightning/pytorch/callbacks/model_checkpoint.py:
```diff
@@ -604,6 +604,7 @@ def _find_last_checkpoints(self, trainer: "pl.Trainer") -> Set[str]:
                 os.path.normpath(p)
                 for p in self._fs.ls(ckpt_path, detail=False)
                 if self.CHECKPOINT_NAME_LAST in os.path.split(p)[1]
+                and os.path.split(p)[1].endswith(self.FILE_EXTENSION)
```
```suggestion
                if f"{self.CHECKPOINT_NAME_LAST}{self.FILE_EXTENSION}" == os.path.split(p)[1]
```

yassersouri replied:
I am not sure if the assumption that the file is always called "last.chk" is correct or not. I thought that it could have other names as well like the epoch could be in the name, etc.

carmocca replied:
@yassersouri are you still interested in finishing this PR?

carmocca reviewed (commented):
Can you add a simple test for the added logic and a [CHANGELOG](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/CHANGELOG.md) entry?

yassersouri commented:
@carmocca Where are the tests for the `model_checkpoint.py`? I couldn't find it in `tests/tests_pytorch
/callbacks/`.

carmocca commented:
They are in `tests/tests_pytorch/checkpointing/test_model_checkpoint.py`

yassersouri commented:
Hi. At the moment I don't have the capacity.

On Mon, Nov 20, 2023 at 7:22 AM Carlos Mocholí ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/lightning/pytorch/callbacks/model_checkpoint.py
> <https://github.com/Lightning-AI/lightning/pull/17072#discussion_r1399360921>
> :
>
> >                  if self.CHECKPOINT_NAME_LAST in os.path.split(p)[1]
> +                and os.path.split(p)[1].endswith(self.FILE_EXTENSION)
>
> @yassersouri <https://github.com/yassersouri> are you still interested in
> finishing this PR?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Lightning-AI/lightning/pull/17072#discussion_r1399360921>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACGSB55HY4ZSSN2K4WKQDDYFNYRZAVCNFSM6AAAAAAVZ4ZX6SVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONBQGA4DAMRYHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
